R/rust_sitrep.R

```diff
@@ -0,0 +1,120 @@
+#' Report on Rust infrastructure
+#'
+#' Prints out a detailed report on the state of Rust infrastructure on the host machine.
+#' @return Nothing
+rust_sitrep <- function() {
+  cargo_v <- get_version("cargo")
+  cargo_msg <- if (is.na(cargo_v)) {
+    c("x" = "{.val cargo}: {.strong not found}")
+  } else {
+    c("v" = "{.val cargo}: {.emph {cargo_v}}")
+  }
+
+  rustup_v <- get_version("rustup")
+  rustup_msg <- if (is.na(rustup_v)) {
+    c("x" = "{.val rustup}: {.strong not found}")
+  } else {
+    c("v" = "{.val rustup}: {.emph {rustup_v}}")
+  }
+
+  msgs <- c(
+    "Rust infrastructure sitrep:",
+    rustup_msg,
+    cargo_msg
+  )
+
+  if (!is.na(rustup_v)) {
+    rustup_status <- rustup_toolchain_target() # nolint: object_usage
+    msgs <- c(
+      msgs,
+      "i" = "host: {.emph {rustup_status$host}}",
+      "i" = "toolchain: {.emph {rustup_status$toolchain}}",
+      "i" = "target{?s}: {.emph {rustup_status$targets}}"
+    )
+  } else {
+    msgs <- c(
+      msgs,
+      "x" = "Cannot determine host, toolchain, and targets without {.val rustup}"
+    )
+  }
+
+  if (is.na(cargo_v)) {
+    msgs <- c(
+      msgs,
+      "x" = "{.val cargo} is required to build {.pkg rextendr}-powered packages"
+    )
+  }
+
+  if (is.na(cargo_v) || is.na(rustup_v)) {
+    msgs <- c(
+      msgs,
+      "i" = "It is recommended to use {.val rustup} to manage {.val cargo} and {.val rustc}",
+      "i" = "See {.url https://rustup.rs/} for installation instructions"
+    )
+  }
+
+  cli::cli_inform(msgs)
+
+  invisible(NULL)
+}
+
+try_exec_cmd <- function(cmd, args = character()) {
+  result <- tryCatch(
+    processx::run(cmd, args, error_on_status = FALSE),
+    error = \(...) list(status = -1)
+  )
+  if (result[["status"]] != 0) {
+    NA_character_
+  } else {
+    result$stdout
+  }
+}
```

Comment:
This is not a big issue, but I believe the following function added in #271 could be replaced by this function.
How about moving this function to utils.R and replacing `cargo_command_available`?

https://github.com/extendr/rextendr/blob/bff80289ee9e4384160dba8a09f1b3ca6ea449ba/R/utils.R#L36-L48

Reply:
Yep, I am looking at it right now. I can do that